### ISSUE TITLE
TPE sampling for integer values

### ISSUE THREAD

**Issue by bouthilx** [Epistimio/orion#600]
When sampling for integer dims in TPE using the CategoricalSampler it seems the number of categories is incorrect. I did not dig yet into the issue to understand why. The problem may be in `TPE._sampe_int_point` since the tests for categorical dimensions are passing. 

Full stack-trace from pytest
```
src/orion/testing/algo.py:396: in test_int_data
    mocker, num, attr, {"x": "uniform(0, 5, discrete=True)"}
src/orion/testing/algo.py:157: in assert_dim_type_supported
    spy = self.spy_phase(mocker, num, algo, attr)
src/orion/testing/algo.py:150: in spy_phase
    self.force_observe(num, algo)
src/orion/testing/algo.py:143: in force_observe
    points = algo.suggest(1)
src/orion/core/worker/primary_algo.py:67: in suggest
    points = self.algorithm.suggest(num)
src/orion/algo/tpe.py:303: in suggest
    candidates = self._suggest_bo(num - len(samples))
<string>:3: in _suggest_bo
    ???
/usr/lib/python3.6/unittest/mock.py:939: in __call__
    return _mock_self._mock_call(*args, **kwargs)
/usr/lib/python3.6/unittest/mock.py:1005: in _mock_call
    result = effect(*args, **kwargs)
/home/bouthilx/.virtualenvs/orion.benchmark/lib/python3.6/site-packages/pytest_mock/plugin.py:97: in wrapper
    r = method(*args, **kwargs)
src/orion/algo/tpe.py:339: in _suggest_bo
    return self._suggest(num, suggest_bo)
src/orion/algo/tpe.py:320: in _suggest
    for candidate in function(num - len(points)):
src/orion/algo/tpe.py:337: in suggest_bo
    return [self._suggest_one_bo() for _ in range(num)]
src/orion/algo/tpe.py:337: in <listcomp>
    return [self._suggest_one_bo() for _ in range(num)]
src/orion/algo/tpe.py:374: in _suggest_one_bo
    self._sample_int_point,
src/orion/algo/tpe.py:419: in sample_one_dimension
    new_point = sampler(dimension, below_points[j], above_points[j])
src/orion/algo/tpe.py:508: in _sample_int_point
    lik_above = sampler_above.get_loglikelis(candidate_points)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <orion.algo.tpe.CategoricalSampler object at 0x7f394dc6d240>, points = array([0, 5, 1, 0])

    def get_loglikelis(self, points):
        """Return the log likelihood for the points"""
>       return numpy.log(numpy.asarray(self.weights)[points])
E       IndexError: index 5 is out of bounds for axis 0 with size 5
```

These are tests incoming in a new Pull Request.

**Comment by donglinjy** [comment]
The root cause of the issue for `uniform(0, 5, discrete=True)` space, when we randomly pick it at the initial phase of tpe, there could be 6 integer from it (0, 1, 2, ..., 5),  but in optimization phase, we treat it as size of 5 ( 0, 1, 2, ..., 4).

**Comment by bouthilx** [comment]
I think both real and integer priors should be inclusive (0, 5). The tricky part is that sampling from uniform to get discrete values requires changing the interval to (0, 6-ish) to get the correct discrete uniform distribution.

**Comment by bouthilx** [comment]
Closed by #619
